src/pkg/provisioner/provisioner.go

```diff
@@ -122,13 +122,38 @@ func cleanup(rootDir, stateDir string) error {
 			return err
 		}
 	}
+	// There are a few files in /var/log that need to exist for daemons to work.
+	// The best way to clear logs is to zero them out instead of deleting them.
 	if err := zeroAllFiles(filepath.Join(rootDir, "var", "log")); err != nil {
 		return err
 	}
 	log.Println("Done cleaning up machine state")
 	return nil
 }
 
+func executeSteps(s *state, c Config) error {
+	for i, step := range c.Steps {
+		// In the case where executeSteps runs after a reboot, we need to skip
+		// through all the steps that have already been completed.
+		if i < s.data.CurrentStep {
+			continue
+		}
+		abstractStep, err := parseStep(step.Type, step.Args)
+		if err != nil {
+			return fmt.Errorf("error parsing step %d: %v", i, err)
+		}
+		if err := abstractStep.run(s); err != nil {
+			return fmt.Errorf("error in step %d: %v", i, err)
+		}
+		// Persist our most recent completed step to disk, so we can resume after a reboot.
+		s.data.CurrentStep++
```

Comment:
should we increment `CurrentStep` after writing to disk?

Reply:
We need to increment CurrentStep first, because we want the new value of CurrentStep to be committed to disk. That way, if the next step reboots (and so doesn't increment CurrentStep and write it to disk), previously executed steps won't be re-run on re-entry.

Reply:
Thanks! Makes sense!